src/transformStateWithClones.js

```diff
@@ -7,7 +7,59 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    stateCopy = { ...stateCopy };
+
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = addProperties(stateCopy, action.extraData);
```

Comment:
The function `addProperties` should not return a new state object. Instead, it should modify the `stateCopy` directly. This avoids the unnecessary creation of an additional object and adheres to the allowed modification of a state copy within the `transformStateWithClones` function.